lib/mongo/server/connection.rb

```diff
@@ -155,7 +155,15 @@ def ping
 
       def deliver(messages)
         write(messages)
-        messages.last.replyable? ? read : nil
+
+        if messages.last.replyable?
+          # Protection against returning the response to a previous request. See
+          # RUBY-1117
+          reply = read until reply && reply.response_to == messages.last.request_id
```

Comment:
Yeah that would improve the average case but it doesn't fix anything, if the first query is slow, when the main thread check out the orphan socket there's nothing to read, so it go ahead, send the second query and reads back the first response.

After thinking about this a bit more with, wouldn't the solution be to close the socket when there's an internal exception before checking it in back in the pool ? So when the thread is killed, the socket would simply be dropped because we don't know in which crappy state it is, the next query will have to reopen it.

Instead of:
```ruby
def with_connection
  connection = checkout
  yield(connection)
ensure
  checkin(connection) if connection
end
```

It would look like:
```ruby
def with_connection
  connection = checkout
  yield(connection)
rescue Exception
  connection.disconnect! if connection
  raise
ensure
  checkin(connection) if connection
end
```

WDYT ?

I also found in the `connection_pool` gem used by everyone that they're using `Thread.handle_interrupt` which seems to help catching interruption like `kill` in https://github.com/mperham/connection_pool/blob/master/lib/connection_pool.rb#L56

I haven't looked much at this and I don't know if it's useful here as the ensure seems to be called properly in our case otherwise the socket would be lost and not reused.

+1 for the op_code check.